File: usecases/src/main/java/com/zeyad/usecases/stores/DataStoreFactory.java
package com.zeyad.usecases.stores;

import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import com.zeyad.usecases.Config;
import com.zeyad.usecases.mapper.DAOMapper;
import com.zeyad.usecases.network.ApiConnection;
import com.zeyad.usecases.utils.DataBaseManagerUtil;
import com.zeyad.usecases.utils.Utils;

public class DataStoreFactory {
    private final static String DB_NOT_ENABLED = "Database not enabled!";
    private static CloudStore mCloudStore;
    private static DiskStore mDiskStore;
    private static MemoryStore mMemoryStore;
    private final DataBaseManagerUtil mDataBaseManagerUtil;
    private final ApiConnection mApiConnection;
    private final DAOMapper mDAOMapper;
    private final boolean withCache;

    public DataStoreFactory(@Nullable DataBaseManagerUtil dataBaseManagerUtil, ApiConnection restApi,
                            DAOMapper daoMapper) {
//        Config.INSTANCE.setHasRealm(dataBaseManagerUtil != null);
//        Config.INSTANCE.setWithSQLite(dataBaseManagerUtil != null);
        mDataBaseManagerUtil = dataBaseManagerUtil;
        mApiConnection = restApi;
        mDAOMapper = daoMapper;
        withCache = Config.INSTANCE.getWithCache();
    }

    /**
     * Create {@link DataStore} .
     */
    @NonNull
    public DataStore dynamically(@NonNull String url, Class dataClass) throws IllegalAccessException {
        if (!url.isEmpty()) {
            return cloud(dataClass);
        } else if (mDataBaseManagerUtil == null) {
            throw new IllegalAccessException(DB_NOT_ENABLED);
        } else {
            return disk(dataClass);
        }
    }

    /**
     * Creates a disk {@link DataStore}.
     */
    public MemoryStore memory() {
        if (withCache && mMemoryStore == null) {
            mMemoryStore = new MemoryStore(Config.INSTANCE.getGson());
        }
        return withCache ? mMemoryStore : null;
    }

    /**
     * Creates a disk {@link DataStore}.
     */
    @NonNull
    public DataStore disk(Class dataClass) throws IllegalAccessException {
        if (!Config.INSTANCE.getWithRealm() || mDataBaseManagerUtil == null) {
            throw new IllegalAccessException(DB_NOT_ENABLED);
        } else if (mDiskStore == null) {
            mDiskStore = new DiskStore(mDataBaseManagerUtil.getDataBaseManager(dataClass), memory());
        }
        return mDiskStore;
    }

    /**
     * Creates a cloud {@link DataStore}.
     */
    @NonNull
    public DataStore cloud(Class dataClass) {
        if (mCloudStore == null) {
            mCloudStore = new CloudStore(mApiConnection,
                    mDataBaseManagerUtil.getDataBaseManager(dataClass), mDAOMapper, memory(),
                    Utils.getInstance());
        }
        return mCloudStore;
    }
}

File: usecases/src/main/java/com/zeyad/usecases/stores/DataStoreFactory.kt
package com.zeyad.usecases.stores

import com.zeyad.usecases.Config
import com.zeyad.usecases.mapper.DAOMapper
import com.zeyad.usecases.network.ApiConnection
import com.zeyad.usecases.utils.DataBaseManagerUtil

class DataStoreFactory(private val dataBaseManagerUtil: DataBaseManagerUtil?,
                       private val restApi: ApiConnection,
                       private val daoMapper: DAOMapper) {

    private val withCache: Boolean = Config.withCache

    fun dynamically(url: String, dataClass: Class<*>): DataStore {
        return if (!url.isEmpty()) {
            cloud(dataClass)
        } else if (dataBaseManagerUtil == null) {
            throw IllegalAccessException(DB_NOT_ENABLED)
        } else {
            disk(dataClass)
        }
    }

    /**
     * Creates a disk [DataStore].
     */
    fun memory(): MemoryStore? {
        if (withCache && memoryStore == null) {
            memoryStore = MemoryStore(Config.gson)
        }
        return if (withCache) memoryStore else null
    }

    /**
     * Creates a disk [DataStore].
     */
    @Throws(IllegalAccessException::class)
    fun disk(dataClass: Class<*>): DataStore {
        if (dataBaseManagerUtil == null) {
            throw IllegalAccessException(DB_NOT_ENABLED)
        } else if (diskStore == null) {
            diskStore = DiskStore(dataBaseManagerUtil.getDataBaseManager(dataClass)!!, memory())
        }
        return diskStore as DiskStore
    }

    /**
     * Creates a cloud [DataStore].
     */
    fun cloud(dataClass: Class<*>): DataStore {
        if (cloudStore == null) {
            cloudStore = CloudStore(restApi,
                    dataBaseManagerUtil?.getDataBaseManager(dataClass)!!, daoMapper, memory())
        }
        return cloudStore as CloudStore
    }

    companion object {
        private var cloudStore: CloudStore? = null
        private var diskStore: DiskStore? = null
        private var memoryStore: MemoryStore? = null
        private const val DB_NOT_ENABLED = "Database not enabled!"
    }
}
